### PR TITLE
Add support for node properties to template

### PR DESCRIPTION
**Pull request by johnl2323**
Add support for node properties (environment variables and tool locations) to AVM templates.  This PR is for this request for enhancement:

Fixes https://github.com/jenkinsci/azure-vm-agents-plugin/issues/566

### Testing done
Updated the BasicConfigAsCodeTests and the basic.yaml to include the nodeProperty jcasc.  Verified that it would read in the jcasc and write out the jcasc.   Test this all with local unit tests.

Manually tested by creating a new template and added new node property env var and verified that it showed up in the Manage Jenkins->Configuration as Code->View Configuration yaml.

This is the first time I am submitting a PR to Jekins and this plugin so let me know if I need to add additional test cases.

I did not run integration tests.

### Submitter checklist
- [X] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your main branch!
- [X] Ensure that the pull request title represents the desired changelog entry
- [X] Please describe what you did
- [X] Link to relevant issues in GitHub or Jira
- [ ] Link to relevant pull requests, esp. upstream and downstream changes
- [X] Ensure you have provided tests - that demonstrates feature works or fixes the issue

### CODE REVIEW

**Comment by timja** [comment]
Could you fix the spotbugs issue please

File: src/main/java/com/microsoft/azure/vmagent/AzureVMAgentTemplate.java [hunk 1]
```diff
@@ -434,6 +441,7 @@ public AzureVMAgentTemplate(
             String agentWorkspace,
             String jvmOptions,
             RetentionStrategy<?> retentionStrategy,
+            List<? extends NodeProperty<?>> nodeProperties,
```
Comment (timja) [inline src/main/java/com/microsoft/azure/vmagent/AzureVMAgentTemplate.java:444]: The constructor should only have new parameters added for mandatory properties, optional values should be added with a setter annotated with `DataBounderSetter` and a getter.

You'll be able to revert the changes to the config.xml then as its a migration testing file

**Comment by johnl2323** [comment]
> Could you fix the spotbugs issue please

Give me a bit to try to figure out what the spotbugs issue is about.   New to Java/Groovy/plugin code.

File: src/main/java/com/microsoft/azure/vmagent/AzureVMAgent.java [hunk 2]
```diff
@@ -181,7 +183,13 @@ public AzureVMAgent(
             String remotingOptions,
             AzureVMAgentTemplate template) throws FormException, IOException {
 
-        super(name, nodeDescription, remoteFS, numExecutors, mode, label, launcher, retentionStrategy, nodeProperties);
+        super(name, remoteFS, launcher);
```
Comment (timja) [inline src/main/java/com/microsoft/azure/vmagent/AzureVMAgent.java:186]: migrated to non deprecated API, required so that I could customise the node properties

File: src/main/java/com/microsoft/azure/vmagent/AzureVMAgent.java [hunk 3]
```diff
@@ -220,6 +228,24 @@ public AzureVMAgent(
...
+        DescribableList<NodeProperty<?>, NodePropertyDescriptor> nodeProperties = template.getNodeProperties();
+        EnvironmentVariablesNodeProperty envVarsNodeProperty = nodeProperties
+                .get(EnvironmentVariablesNodeProperty.class);
+
+        List<EnvironmentVariablesNodeProperty.Entry> newEntries = new ArrayList<>();
+        if (envVarsNodeProperty != null) {
+            nodeProperties.remove(envVarsNodeProperty);
+            newEntries.addAll(envVarsNodeProperty.getEnv());
+        }
+        newEntries.add(new EnvironmentVariablesNodeProperty.Entry("FQDN", fqdn));
```
Comment (timja) [inline src/main/java/com/microsoft/azure/vmagent/AzureVMAgent.java:243]: merges the one this plugin previously set with the ones potentially in the template

File: src/main/java/com/microsoft/azure/vmagent/AzureVMAgentTemplate.java [hunk 4]
```diff
@@ -384,6 +389,10 @@ public int hashCode() {
 
     private String licenseType;
 
+    @SuppressFBWarnings(value = "SE_BAD_FIELD", justification = "They are serializable in practice")
```
Comment (timja) [inline src/main/java/com/microsoft/azure/vmagent/AzureVMAgentTemplate.java:392]: Not 100% confident on this, but I checked all the node properties in core and they were all serialisable.

This will be popping up because the plugin must serialise this class when its keeping track of deployments.
I think this will be ok and in my testing I didn't see any issues with it

File: src/main/java/com/microsoft/azure/vmagent/AzureVMAgentTemplate.java [hunk 5]
```diff
@@ -827,6 +848,10 @@ private Object readResolve() {
             storageAccountType = SkuName.STANDARD_LRS.toString();
         }
 
+        if (nodeProperties == null) {
+            nodeProperties = new DescribableList<>(Saveable.NOOP);
```
Comment (timja) [inline src/main/java/com/microsoft/azure/vmagent/AzureVMAgentTemplate.java:852]: I noticed a null pointer when it was being deserialised and no node properties had been set, this will work around that

File: src/main/resources/com/microsoft/azure/vmagent/AzureVMComputer/configure.jelly [hunk 6]
```diff
@@ -3,22 +3,23 @@
   <l:layout permission="${it.EXTENDED_READ}" title="${it.name} Configuration">
     <st:include page="sidepanel.jelly"/>
     <l:main-panel>
+      <j:set var="instance" value="${it.node}" />
+      <j:set var="descriptor" value="${instance.descriptor}" />
+      <j:set var="readOnlyMode" value="true" />
```
Comment (timja) [inline src/main/resources/com/microsoft/azure/vmagent/AzureVMComputer/configure.jelly:8]: cleans up the UI by using read only controls for everything on the page.

In the earlier version the node properties appeared to be modifiable but we don't want that

File: src/main/resources/com/microsoft/azure/vmagent/AzureVMComputer/configure.jelly [hunk 7]
```diff
@@ -3,22 +3,23 @@
...
       <l:app-bar title="${it.name} Configuration" />
-      <f:entry title="${%Name}">
-        <f:readOnlyTextbox value="${it.name}"/>
-      </f:entry>
+      <p><a href="${it.azurePortalLink}">View on Azure Portal</a></p>
```
Comment (timja) [inline src/main/resources/com/microsoft/azure/vmagent/AzureVMComputer/configure.jelly:10]: With the node properties this was pushed too far down, moving to the top keeps it easily discoverable

File: src/main/resources/com/microsoft/azure/vmagent/AzureVMComputer/configure.jelly [hunk 8]
```diff
@@ -3,22 +3,23 @@
   <l:layout permission="${it.EXTENDED_READ}" title="${it.name} Configuration">
     <st:include page="sidepanel.jelly"/>
     <l:main-panel>
+      <j:set var="instance" value="${it.node}" />
+      <j:set var="descriptor" value="${instance.descriptor}" />
+      <j:set var="readOnlyMode" value="true" />
       <l:app-bar title="${it.name} Configuration" />
-      <f:entry title="${%Name}">
```
Comment (timja) [inline src/main/resources/com/microsoft/azure/vmagent/AzureVMComputer/configure.jelly:7]: not needed as the title has the name

**Comment by timja** [comment]
Hi, I tested this and it wasn't working to actually add the node properties, I've fixed that and tidied it up a bit.

Thanks for the contribution!
